refactor: change to the official riscv32emac toolchain

jarkkojs opened the pull request:
Change from the Parity's rv32e fork to the official rustc toolchain.

Makes sure that also PolkaVM exports get relocation entry with the '--relocatable' parameters. Otherwise the linker will only add the relocs, whic are referenced by the code directly.

Flag DWARF processing under the feature flag 'dwarf' for the moment.

Link: https://github.com/paritytech/rustc-rv32e-toolchain/
Closes: #178

athei commented on crates/polkavm-linker/src/program_from_elf.rs:
```diff
@@ -8311,8 +8311,9 @@ where
         let mut string_cache = crate::utils::StringCache::default();
-        let dwarf_info = crate::dwarf::load_dwarf(&mut string_cache, &elf, &relocations)?;
-        location_map = dwarf_info.location_map;
+        if cfg!(feature = "dwarf") {
+            location_map = (crate::dwarf::load(&mut string_cache, &elf, &relocations)?).0;
+        }
```
Why would it be a feature flag? Let's just make it a runtime option.

jarkkojs replied:
Sorry, define "runtime option".

koute commented on guest-programs/build-benchmarks.sh:
```diff
@@ -1,14 +1,10 @@
 set -euo pipefail
 cd -- "$(dirname -- "${BASH_SOURCE[0]}")"
 
 source ../ci/jobs/detect-or-install-riscv-toolchain.sh
```
This line should be deleted, and that script should be deleted altogether.

jarkkojs replied:
`#!/usr/bin/env bash` is POSIX cross-compatible, and not just in theory, there is variance with `/bin/bash` and `/usr/bin/bash`. Technically (if script was kept) this should be a separate fix.

But since your request was to remove this, and the issue defined running this script as the test case. How do I test it afterwards?

jarkkojs replied:
So I keep this open until the pull request is final. I updated the PR because it already fixes two issues in DWARF code.

Fixed: overlap bug
Open: how to deal DW_AT_specification with zero offset

koute replied:
> `#!/usr/bin/env bash` is POSIX cross-compatible, and not just in theory, there is variance with `/bin/bash` and `/usr/bin/bash`. Technically (if script was kept) this should be a separate fix.

This wasn't the line I was talking about. (: In general, I don't care which one it is, as long as it works. I always tend to use `/bin/bash` by default because that seems to work everywhere for me, but if you use some weird distro where it doesn't then feel free to go wild and change it.

I highlighed this line (the line immediately above my comment):

```
source ../ci/jobs/detect-or-install-riscv-toolchain.sh
```

If we're switching to the upstream toolchain then we don't need this line, so it should be deleted along with the script.

koute commented on crates/polkavm-linker/Cargo.toml:
```diff
@@ -1,3 +1,7 @@
+[features]
+default = []
+dwarf = []
```
We shouldn't have any features here (DWARF support should be always compiled in), and we already have a runtime flag which makes the linker ignore DWARF (the `--strip` flag in polkatool/`set_strip` in linker's config struct)

jarkkojs replied:
So in today's meeting we made the plan that since DWARF parsing is a separate issue from relocation, and it would be better to get the official toolchain in use, it is flagged and a new issue is created to fix it.

jarkkojs commented on crates/polkavm-linker/Cargo.toml:
```diff
@@ -1,3 +1,7 @@
+[features]
+default = []
```
By evolution and re-learning Rust, i.e. I tried to recall how feature flags where defined and this is only cruft. I don't mind removing it.

jarkkojs commented on guest-programs/.cargo/config.toml:
DOne

jarkkojs commented:
Not worth commenting just yet. There's still a few issues but all are related just twists in DWARF data, which did not exist when compiled with Parity's toolchain.

Summary:
- Most of the fixes already in place might need some tweaking but mostly are correctly identified bugs.
- There's one issue at least left that I simply did not yet investigate properly but can be addressed probably just by looking the binary with rz-bin.
- @athei asked to remove .cargo/config.toml but I'm not sure about this but open for feedback. @koute on the other asked to remove build-bechmark.sh. So some *overall* conclusions and consensus "do this with build shenanigans" would be nice here. No focus on this before I've fixed the remaining DWARF issues.

So I do a temporary work branch to finalize this pull request, thus does not yet much sense to give any detailed feedback on the code itself. I've tried to add some rz-bin output to each commit so that it describes clearly the situation in the binary and kind of helps to understand from those grounds what the code change does (as otherwise these make no sense at all).

jarkkojs commented:
So this iteration #2 of the pull request.

Open:

```
.github/workflows/rust.yml:      run: ./ci/jobs/detect-or-install-riscv-toolchain.sh
.github/workflows/rust.yml:      run: ./ci/jobs/detect-or-install-riscv-toolchain.sh
ci/jobs/build-guests.sh:source ./ci/jobs/detect-or-install-riscv-toolchain.sh
ci/jobs/clippy.sh:source ./ci/jobs/detect-or-install-riscv-toolchain.sh
ci/jobs/rustfmt.sh:source ./ci/jobs/detect-or-install-riscv-toolchain.sh
ci/run-all-tests.sh:source ./ci/jobs/detect-or-install-riscv-toolchain.sh
guest-programs/build-examples.sh:source ../ci/jobs/detect-or-install-riscv-toolchain.sh
```

I think it is better to have a checkpoint (feedback) first before fixing up these (in a separate commit).

Changes to `dwarfs.rs` are bottom because I wanted to test them also with the previous toolchain. With these changes now also `build-benchmark.sh` passes with the new toolchain.

koute commented on crates/polkavm-linker/src/dwarf.rs:
```diff
@@ -1455,9 +1466,7 @@ where
-                    return Err(ProgramFromElfError::other(
-                        "failed to process DWARF: subprogram has a decl location but no name",
-                    ));
+                    log::warn!("failed to process DWARF: subprogram has a decl location but no name");
```
Hm, so do you know if these are due to broken DWARF, or if this is just "normal"? If it's normal then this log should probably be downgraded to `debug` or `trace`.

But if you don't know then it's fine as-is.

(This applies to every `warn!` you've added here.)

jarkkojs replied:
We need to figure these out (as discussed) in use. Right now I think this right action to take. Features and bugs really make them resolve themselves.

In the long-run I'd also consider a split to "subprogram" and "subroutine" phases as subprograms can be parsed without subroutines and with experimental change this really seems to make the whole so much more clean.

Subprograms alone have their own special cases such as out-of-line inlines (which are not actually inlines) but they never have subroutine parents. Only subroutines point to the subprograms. I think this would help to sort out many of these. I did not want to include this change just yet because it goes way beyond toolchain change.

It's like the next thing when something major specific to DWARF needs to be done. So not dealing with all of this right now (the ones that made sense to deal are separate commits) but I have ideas and plans how to deal with them when we need to :-) 

So with this I think we can resolve this (in the context of this pull request).

koute reviewed (commented):
Looks mostly good to me!

We can merge this once we fix clippy.

jarkkojs commented:
So I there was issue I was looking into on removing that CI script see first this temp diff:

```
diff --git a/crates/polkavm-linker/src/program_from_elf.rs b/crates/polkavm-linker/src/program_from_elf.rs
index 0c60355..d18dcff 100644
--- a/crates/polkavm-linker/src/program_from_elf.rs
+++ b/crates/polkavm-linker/src/program_from_elf.rs
@@ -1408,6 +1408,7 @@ where
     let _ = b.read(target.offset as usize)?;
 
     let version = b.read_byte()?;
+    log::debug!("section {} {}", section.name(), version);
     if version != 1 && version != 2 {
         return Err(format!("unsupported extern metadata version: '{version}' (expected '1' or '2')"));
     }
diff --git a/guest-programs/build-examples.sh b/guest-programs/build-examples.sh
index 39542a3..0e00312 100755
--- a/guest-programs/build-examples.sh
+++ b/guest-programs/build-examples.sh
@@ -12,13 +12,22 @@ fi
 
 function build_example () {
     output_path="output/$1.polkavm"
-    current_dir=$(pwd)
 
     echo "> Building: '$1' (-> $output_path)"
-    RUSTFLAGS="-C target-feature=+c -C relocation-model=pie -C link-arg=--emit-relocs -C link-arg=--unique --remap-path-prefix=$(pwd)= --remap-path-prefix=$HOME=~" cargo build -q --release --bin $1 -p $1
-    cd ..
-    cargo run -q -p polkatool link --run-only-if-newer -s guest-programs/target/riscv32ema-unknown-none-elf/release/$1 -o guest-programs/$output_path
-    cd $current_dir
+
+    RUSTFLAGS="--remap-path-prefix=$(pwd)= --remap-path-prefix=$HOME=~" \
+    cargo build  \
+        -Z build-std=core,alloc \
+        --target "$PWD/riscv32emac-unknown-none-polkavm.json" \
+        -q --release --bin $1 -p $1
+
+    pushd ..
+
+    cargo run -q -p polkatool link \
+        --run-only-if-newer guest-programs/target/riscv32emac-unknown-none-polkavm/release/$1 \
+        -o guest-programs/$output_path
+
+    popd
 }
 
 build_example "example-hello-world"
```

The result:

```
❯ RUST_BACKTRACE=1 RUST_LOG="polkavm_linker::program_from_elf=debug" guest-programs/build-examples.sh   
> Building: 'example-hello-world' (-> output/example-hello-world.polkavm)
~/work/github.com/jarkko/polkavm ~/work/github.com/jarkko/polkavm/guest-programs
[DEBUG polkavm_linker::program_from_elf] section .polkavm_metadata 1
[DEBUG polkavm_linker::program_from_elf] section .polkavm_metadata 0
ERROR: failed to link "guest-programs/target/riscv32emac-unknown-none-polkavm/release/example-hello-world": failed to parse extern metadata: unsupported extern metadata version: '0' (expected '1' or '2')
```

I sanity checked again with old toolchain that my DWARF changes could not possibly cause any of this and they do not. `build-examples.sh` works perfectly with all of them. I also checked what happens with benchmarks and with them the version is 1 for both exports.

I looked at also `polkavm-derive` crate and found the generators and it version 1 is defined for `METADATA`.

I keep also Friday off work (like I did Wed) to keep my hours in balance so looking into this next Monday.

PS. A more standard convention would to use ELF note section (e.g. `.note.polkavm`) for this. Just saying this in completeness no strong push ATM :-)

jarkkojs commented:
I'll check on Monday what the binary dumps comparing exports section (or whatever it was called) with old/new toolchain (dunno why did not came to mind Thu).

athei commented on guest-programs/build-examples.sh:
```diff
@@ -5,20 +5,26 @@ cd -- "$(dirname -- "${BASH_SOURCE[0]}")"
-    cd $current_dir
+
+    local target="riscv32emac-unknown-none-polkavm"
+    local extra_flags="-C target-feature=+c -C relocation-model=pie -C link-arg=--emit-relocs -C link-arg=--unique --remap-path-prefix=$(pwd)= --remap-path-prefix=$HOME=~"
```
Do we still need all of those args? They seem to be already set by the target.

jarkkojs replied:
Probably not, but did not yet remove it given inconsistent data generated to the exports section. Consider this commit WiP up until the issue root caused. I'll keep this open tho.

jarkkojs replied:
Let's enumerate and evaluate:

1. `-C target-feature=+c`
2. `-C relocation-model=pie`
3. `-C link-arg=--emit-relocs`
4. `-C link-arg=--unique`
5. `--remap-path-prefix=$(pwd)= --remap-path-prefix=$HOME=`

2, 3 and 4 are dead obvious removal but I'm lost with 1 and 5. I don't even know what they are, and what they were meant to do originally. Perhaps @koute can bring context here? Would be educational to know even they would get scraped :-)

I'm also quite confused, informal googling gives me the idea that 5 should be in fact `--remap-path-prefix-from=$(pwd)= --remap-path-prefix-to=$HOME=`. Still, causes no compilation errors.

koute replied:
(1) is enabling the C extension because the original toolchain didn't have it enabled by default; on the upstream toolchain that should be unnecessary (because the target JSON file should have it enabled by default).

(5) replaces paths in the debug info; by default you'd get paths like `/home/user/.cargo/...` hardcoding the username of whoever compiled the file and `/path/to/checkout/polkavm` hardcoding the exact place where you've checked out the repo, and we don't want that (ideally we want builds to be bit-by-bit reproducible), so this snips these paths and makes them relative

athei commented on guest-programs/build-examples.sh:
```diff
@@ -5,20 +5,26 @@ cd -- "$(dirname -- "${BASH_SOURCE[0]}")"
+    pushd ..
+
+    cargo run -q -p polkatool link \
+        --run-only-if-newer -s guest-programs/target/$target/release/$1 \
```
`-s` means we are stripping and hence skipping DWARF in the linker. Is that intended?

jarkkojs replied:
Probably/possibly not. I'll check this. I had no idea what the option is doing and it existed before the change.

To be more formal, let's compare it to `build-benchmark.sh`:

```
    cargo run -q -p polkatool link \
        --run-only-if-newer guest-programs/target/riscv32emac-unknown-none-polkavm/release/$1 \
        -o guest-programs/target/riscv32emac-unknown-none-polkavm/release/$1.polkavm
```

Yep, it's likely that it is not needed.

koute replied:
This is intended here in this script because the resulting blob is checked into the repo.

jarkkojs commented:
zZz and wait for input... I wrote on Sunday (and sort of used half of my Monday, also rebased the patch set and fixed merge conflicts and hit my head wall with this issue under discussion) extra patch that simply allows to optionally pick benchmark with a command-line parameter. It's easier to fix build and linking issues because you don't have to comment out stuff by hand from `build-benchmarks.sh`.

jarkkojs commented:
After fixing up the parameters in `build-examples.sh` according to the comments:

```bash
❯ guest-programs/build-examples.sh 
> Building: 'example-hello-world' (-> output/example-hello-world.polkavm)
~/work/github.com/jarkko/polkavm ~/work/github.com/jarkko/polkavm/guest-programs
ERROR: failed to link "guest-programs/target/riscv32emac-unknown-none-polkavm/release/example-hello-world": failed to parse extern metadata: unsupported extern metadata version: '0' (expected '1' or '2')
```

So the problem still persists.

jarkkojs commented:
@koute here's the dumps.

New toolchain:

```bash
 objdump -s -j .polkavm_exports ./guest-programs/target/riscv32emac-unknown-none-polkavm/release/example-hello-world

./guest-programs/target/riscv32emac-unknown-none-polkavm/release/example-hello-world:     file format elf32-little

Contents of section .polkavm_exports:
 0000 01000000 00000000 00                 .........       
```

Old toolchain:

```bash
❯ objdump -s -j .polkavm_exports ./guest-programs/target/riscv32ema-unknown-none-elf/release/example-hello-world 

./guest-programs/target/riscv32ema-unknown-none-elf/release/example-hello-world:     file format elf32-little

Contents of section .polkavm_exports:
 0000 01942101 00641101 00                 ..!..d...       
```

jarkkojs commented:
Just documenting RISC-V relocations.

Both truths hold also for RISC-V when it comes to relocations:

1. `R_RISCV_32`: **S +A** (also known as *absolute relocation*)
2. `R_RISCV_PC32`: **S+A-P** (also known as *relative relocation*)

In the latter **P** is location of the instruction/whatever in the memory which is computed by the formula `sh_addr + r_offset`. This is 1:1 exact match how 386 relocations work in ELF.

Relocations are not universally independent on section addresses. Only the absolute relocations are like this. This good to keep in mind and also that it depends case by case.

jarkkojs commented:
Follow-up: https://github.com/paritytech/polkavm/issues/199

jarkkojs commented:
Looking this next time on Monday, if it is still open but I'd hope that since I created follow-up issues and this passes the CI, it'd be a great time to merge it, in order to get a milestone (and move forward with #199).

jarkkojs commented:
Please use "rebase and merge" instead of "squash and merge" so that my splits don't go wasted, and will in future useful for `git bisect` and `git blame -s` when seeking for failing commits. The default was "squash and merge", thus the remark.

koute commented:
> Please use "rebase and merge" instead of "squash and merge" so that my splits don't go wasted

If the commits are nice I always use rebase and merge (and in general I personally prefer a rebase-based workflow myself), but FYI this is not true for Parity at large (if you're going to contribute to `polkadot-sdk` then there the policy is to never rebase, and to always squash).

koute commented on ci/jobs/build-guests.sh:
```diff
@@ -1,13 +1,11 @@
 set -euo pipefail
-cd -- "$(dirname -- "${BASH_SOURCE[0]}")"
+
+cd "${0%/*}/"
```
Is there actually a difference between what I've been doing previously and this approach? Is this one superior?

jarkkojs replied:
It is heck a lot more cleaner and compact at least.

jarkkojs replied:
It's not that different to Rust where you can do bunch of stuff with `{}`. I'd recommend to take these into use :-) The logic is not complex once you know them

xermicus replied:
Does this work fine with zsh too? Bash only features will make Mac userss.

jarkkojs replied:
It's formally same as asking does this work on a Python interpreter too. Why do you find difficult to install bash? I use zsh too.

Zsh has the construct I described, if that was the question.

jarkkojs replied:
For what is worth, I checked running `zsh polkavm/guest-programs/build-examples.sh`, so if you really want to go that route, it is possible. Zsh as a scripting language is by a large part superset of bash.

koute commented on guest-programs/build-benchmarks.sh:
```diff
@@ -71,17 +65,28 @@ if [ "${SOLANA_PLATFORM_TOOLS_DIR:-}" == "" ]; then
+        --target "$PWD/riscv32emac-unknown-none-polkavm.json" \
+        -q --release --bin $1 -p $1
+
+    pushd ..
```
Can you silence the `pushd`/`popd`? It spams the console with the paths, which is just unnecessary noise.

jarkkojs replied:
+1

I'll look into this on Monday.

jarkkojs replied:
I fixed up this one.

koute commented on crates/polkavm-linker/src/program_from_elf.rs:
```diff
@@ -1410,9 +1414,18 @@ where
-    let version = b.read_byte()?;
+    let mut version = b.read_byte()?;
+
+    // FIXME: `.polkavm_metadata` section is padded with 0 when it starts with
+    // an import for unknown reason.
+    if version == 0 {
+        version = b.read_byte()?;
+    }
```
This should be deleted.

The issue seems to be due to me not specifying a proper repr on the metadata structs; here's a patch with a fix:

```diff
diff --git a/crates/polkavm-derive-impl/src/abi_support_impl/common_private.rs b/crates/polkavm-derive-impl/src/abi_support_impl/common_private.rs
index 6c06da6..9dbd7aa 100644
--- a/crates/polkavm-derive-impl/src/abi_support_impl/common_private.rs
+++ b/crates/polkavm-derive-impl/src/abi_support_impl/common_private.rs
@@ -395,7 +395,7 @@ impl PackReturnTy for (Reg, Reg) {
 pub struct MetadataPointer(pub *const u8);
 unsafe impl Sync for MetadataPointer {}
 
-#[repr(packed)]
+#[repr(C, packed(1))]
 pub struct ExternMetadataV1 {
     pub version: u8,
     pub flags: u32,
@@ -405,7 +405,7 @@ pub struct ExternMetadataV1 {
     pub output_regs: u8,
 }
 
-#[repr(packed)]
+#[repr(C, packed(1))]
 pub struct ExternMetadataV2 {
     pub version: u8,
     pub flags: u32,
```

jarkkojs replied:
I agree but since it took me so long time to find out the bug, and find some way to address it, I'd see it would remove from us valuable information to remove my workaround. Especially since the workaround has been marked with `FIXME`, i.e. everyone can see it was not final, and since I already created #199.

So rather than overwriting workaround (while not disagreeing with the actual fix) I'm open here for two alternatives:

1. I'll add a fix on top of this closing #199.
2. We'll finalize this PR and then close #199.

I.e. I see my workaround and your final fix both valuable as "documentation", and acting otherwise would erase useful historical data.

jarkkojs replied:
Since I have the new bug to make the fix already opened, I linked the suggestion there, and will resolve this.

koute commented on guest-programs/bench-prime-sieve/src/unrolled.rs:
```diff
@@ -91,6 +91,7 @@ fn reinterpret_slice_mut_u64_u8(words: &mut [u64]) -> &mut [u8] {
 /// Storage structure implementing standard linear bit storage, but with a hybrid bit setting strategy:
 /// - dense resetting for small skip factors
 /// - sparse resetting for larger skip factors
+///
```
Is this due to `cargo fmt`?

jarkkojs replied:
Not that one. These were reported by the clippy script.

jarkkojs replied:
I did the absolute minimal to keep that script happy.

jarkkojs commented:
> > Please use "rebase and merge" instead of "squash and merge" so that my splits don't go wasted
> 
> If the commits are nice I always use rebase and merge (and in general I personally prefer a rebase-based workflow myself), but FYI this is not true for Parity at large (if you're going to contribute to `polkadot-sdk` then there the policy is to never rebase, and to always squash).

OK cool it was only what Github seems to recommend me as default (again apparently) :-)

jarkkojs commented:
I'll revisit the remaining issues on Monday. Took 4 weeks in the end to get here but zero actual technical debt generated (maybe even a bit reduced in details). It somehow exploded from that once additional `ld` parameter I added to the original JSON (`--relocatable`).

jarkkojs commented:
Also I think Monday would be good time because then the response time any possible issues is fastest (next week I'll be available full 5 days).

jarkkojs commented:
I addressed the reported issues so I'll merge this (including #199).
